### PR TITLE
Abstract active layer evaluation away from the Widget Filter

### PR DESCRIPTION
**Pull request by paynecrl97**
This PR was originally https://github.com/OrchardCMS/Orchard/pull/5363, but I created this new PR in order to merge into `dev` as opposed to `1.9.x`

### CODE REVIEW

File: src/Orchard.Web/Modules/Orchard.Widgets/Services/DefaultLayerEvaluationService.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+            Logger = NullLogger.Instance;
+            T = NullLocalizer.Instance;
+
+            _activeLayerIDs = new LazyField<IEnumerable<int>>();
+            _activeLayerIDs.Loader(PopulateActiveLayers);
+        }
+
+        public ILogger Logger { get; set; }
+        public Localizer T { get; private set; }
```
Comment (DanielStolt) [inline src/Orchard.Web/Modules/Orchard.Widgets/Services/DefaultLayerEvaluationService.cs]: This one should have a public setter.

File: src/Orchard.Web/Modules/Orchard.Widgets/Services/DefaultLayerEvaluationService.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+                    if (_ruleManager.Matches(activeLayer.LayerRule)) {
+                        activeLayerIds.Add(activeLayer.ContentItem.Id);
+                    }
+                }
+                catch (Exception e) {
+                    Logger.Warning(e, T("An error occurred during layer evaluation on: {0}", activeLayer.Name).Text);
+                }
+            }
+
+            return activeLayerIds;
```
Comment (DanielStolt) [inline src/Orchard.Web/Modules/Orchard.Widgets/Services/DefaultLayerEvaluationService.cs]: ... and instead do the `.ToArray()` here.

File: src/Orchard.Web/Modules/Orchard.Widgets/Filters/WidgetFilter.cs [hunk 3]
```diff
@@ -51,25 +49,7 @@ public class WidgetFilter : FilterProvider, IResultFilter {
...
-                        activeLayerIds.Add(activeLayer.ContentItem.Id);
-                    }
-                }
-                catch(Exception e) {
-                    Logger.Warning(e, T("An error occured during layer evaluation on: {0}", activeLayer.Name).Text);
-                }
-            }
-
-            IEnumerable<WidgetPart> widgetParts = _widgetsService.GetWidgets(layerIds: activeLayerIds.ToArray());
+            var widgetParts = _widgetsService.GetWidgets(_layerEvaluationService.GetActiveLayerIds().ToArray());
```
Comment (DanielStolt) [inline src/Orchard.Web/Modules/Orchard.Widgets/Filters/WidgetFilter.cs]: Personal preference maybe but I think it's nicer to maintain a pattern across the codebase where consumers of the API to continue to treat the `IEnumerable<T>` as `IEnumerable<T>` and trust the API to return the simplest underlying form possible. So in this case, trust LayerEvaluationService to return an array... [and see below]

File: src/Orchard.Web/Modules/Orchard.Widgets/Services/DefaultLayerEvaluationService.cs (hunk 1 above)
Comment (paynecrl97) [inline src/Orchard.Web/Modules/Orchard.Widgets/Services/DefaultLayerEvaluationService.cs]: It should. I'm not sure how I've done that :confused:

**Pull request merged**